### PR TITLE
jsm lab starter

### PR DESCRIPTION
**Pull request by jsm13**
The lab-starter branch is the same as 2_active_record. The changes in this PR add instructions for the lab to the Part 3 of the readme and adds starter code which is about what has been done in class over the last couple days. 

I'm continuing working but if I could get some eyes and comments on this I'd really appreciate it
@jshawl @adambray @andrewsunglaekim @mattscilipoti @amaseda  @RobertAKARobin @nolds9 @ebirving

### CODE REVIEW

File: readme.md [hunk 1]
```diff
@@ -101,3 +101,176 @@ In the console...
...
 
 ## Part 3 - Build a RESTful Interface using Sinatra
+
+First you wrote the schema for a database to store records of artists and their songs for a music catalog app. You then seeded the database with sample data and we were able to interact with that data (performing CRUD operations) using SQL queries in Postgresql. Next, you connected ActiveRecord to the database which enhanced your ability to interact with the data in a Ruby environment. Still, our only way to interact with the data is in a console.
+
+Today, you are going to use Sinatra to listen for HTTP requests and serve HTML (rendered from ERB templates) in response.
+
+### Part 3.1: Root Route, Home View, and Layout View
+
+1. Run your application: `ruby app.rb` and go to localhost:4567 in your browser... Sinatra doesn't know this ditty.
```
Comment (jshawl) [inline readme.md:111]: is there a cloneable URL for people who don't have a working `app.rb`?

File: readme.md [hunk 2]
```diff
@@ -101,3 +101,176 @@ In the console...
...
+- The erb method will also check for a layout.erb file where you define the part of the template shared by all the pages in our app (e.g. head/body elements, header/footer, navigation bar, etc). In layout.erb need to include a line of embedded ruby with the keyword `<%= yield %>`. This tells the method to render the named template (`:home`) at this point in the layout.erb template. In the below steps, regularly check out localhost:4567 and consider the effects of your code changes
+
+    - Make a layout.erb file in views/
+
+    - As always, start simple, maybe just a heading. Check it out in the browser
+
+    - Then expand adding head and body elements, layout.erb should produce well formed HTML and since there is no embedded ruby in it yet, it should be well formed HTML
+
+    - Create a __public/__ directory in the root of your project. By default, Sinatra looks for static assets in a directory named __public/__ in the root directory of the project. In the __public/__ create a stylesheet.
+
```
Comment (jshawl) [inline readme.md:149]: maybe hint here about how to link to files in the public folder, i.e. dont include the word `public`

File: readme.md [hunk 3]
```diff
@@ -101,3 +101,176 @@ In the console...
...
+    - Back in layout.erb add a link to your stylesheet in the head. In the body add a nav element containing links to '/artists' and '/songs'. Finally below the nav element, add an embedded ruby expression calling `yield`.
+
+    - Take a moment using your browser's dev tools to look at how the HTML rendered in from home.erb nests into the HTML rendered from layout.erb where we included `<%= yield %>`.
+
+    - Spend a few minutes making this page look nice/pleasant -- as you keep working on the app, you're going to see it a whole bunch
+
+### Part 3.2: Artist Controller and Views
+
+1. Make a __controllers/__ directory in __tunr_sinatra/__ and create a file __artists.rb__ which will hold the routes for your artists
+
```
Comment (jshawl) [inline readme.md:159]: is controllers a new term? maybe give a brief definition: it's the c in MVC. We'll organize our routes and their actions by controllers

File: readme.md [hunk 4]
```diff
@@ -101,3 +101,176 @@ In the console...
...
+    6. Update (submit form to update existing artist) - PUT "/artists/:id"
+
+    7. Destroy (delete an existing artist) - DELETE "/artists/:id"
+
+- Make your __Index__ route for artists
+
+  - In the browser, try to follow the link to "/artists" in your nav bar. You see Sinatra doesn't know this ditty. Not shocking as you have not yet defined it.
+
+  - While it is possible to put the "/artists" route in app.rb, it helps keeps thing organized to keep a separate file where we define all of our artist routes and then use `require_relative` to import that file. These kinds of files which handle interaction between the our models and our views are our controllers so make a __controllers/__ directory and create a file artists.rb. Put in the route Sinatra suggests
```
Comment (jshawl) [inline readme.md:182]: maybe name the file `artists_controller.rb` instead

File: readme.md [hunk 5]
```diff
@@ -101,3 +101,176 @@ In the console...
...
+
+- Now in the controller you want to get data from the database (by means of ActiveRecord and our models) and use it to render the views (using erb).
+
+  - In your GET /artists route add binding.pry.
+    ```ruby
+    get "/artists" do
+      binding.pry
+      erb :"artists/index"
+    end
+    ```
```
Comment (jshawl) [inline readme.md:214]: and view the console when you reload the `/artists` page

File: readme.md [hunk 6]
```diff
@@ -101,3 +101,176 @@ In the console...
...
+
+  - Use your Artist model to get the array of all artists (like in [Part 2.3](##part-2-3) above).
+
+  - In the artist controller, where you had binding.pry, declare an instance variable referencing the array of all artists so [it can be used  in your view](https://github.com/ga-dc/curriculum/blob/master/04-ruby-mvc-sinatra/sinatra-rest/views.md#passing-variables-to-views).
+
```
Comment (jshawl) [inline readme.md:219]: :+1:

File: readme.md [hunk 7]
```diff
@@ -101,3 +101,176 @@ In the console...
...
+    - __Delete__ an artist? Like edit, this also makes sense on the page of a particular artist
+
+  - In views/artists/index.erb template a page which uses the artists array to list all of the artists and includes a link to the new artist form page.
+
+- __Creating__ new artists
+
+  - Trying to follow the links to the new artist page will lead to the same "Sinatra doesn't know this ditty" error.
+
+  - In the artist controller, define a route for GET "artists/new" and create your new page: views/artists/new.erb. The new page should have a [form which makes a POST request](https://github.com/ga-dc/curriculum/blob/master/04-ruby-mvc-sinatra/sinatra-rest/rest.md#post) to "/artists"
+
```
Comment (jshawl) [inline readme.md:249]: really diggin these links back to curriculum :star:

File: readme.md [hunk 8]
```diff
@@ -101,3 +101,176 @@ In the console...
...
+- __Showing__ an artist
+
+  - Thinking back to our artist index page, a plain text list of the artists in the database isn't super useful, even with the very impressive ability to add to that list. What we would like to do is give each individual artist a page. ERB is going to be an enormous asset here. You can define a generic template for showing an artist and then provide the view an artist instance with which to render the form.
+
+    - The challenge here is to [take information from](https://github.com/ga-dc/curriculum/tree/master/04-ruby-mvc-sinatra/sinatra-rest#the-params-hash) the GET artists/:id and use that with ActiveRecord to retrieve the artist instance requested. Spend a little time in pry playing with these ideas.
+
+    - HTML forms only support GET and POST methods so we need to use a work-around to use our PUT and DELETE methods we need for updating and deleting artists. We create a form with `action='/artist/<id here>' method='put'` and any form elements we need but the first child of the form should be `<input type="hidden" name="_method" value="put">` or `value="delete"` for the delete route
+
```
Comment (jshawl) [inline readme.md:264]: recommend a full example:

``` html
<form method="post" action="/artists/3">
  <input type="hidden" name="_method" value="put">
</form>
```

File: readme.md [hunk 9]
```diff
@@ -101,3 +101,176 @@ In the console...
...
+  - Using the techniques covered so far, you have the means implement the 7 RESTful routes discussed above for artists  and create the views for your routes to respond with (giving your user full CRUD functionality over your database!).
+
+### Part 3.3 (Bonus)
+
+- If you are comfortable, begin implementing routes for songs -- consider carefully how a user wants to access songs. Remember that CRUD and ISNECUD are more ideas than hard and fast rules and we structure our app in a way that makes sense to how it will be used.
+
+### Part 3.4 (Super bonus)
+
+- Add a whole new model of playlists to the app
```
Comment (jshawl) [inline readme.md]: third

File: readme.md [hunk 10]
```diff
@@ -101,3 +101,176 @@ In the console...
...
+
+### Part 3.5 (You're either time-traveling or not sleeping bonus)
+
+- Add users and account management
```
Comment (jshawl) [inline readme.md:282]: #### Where to go next

Check out [HAML](http://haml.info/)! Convert your ERB views to use HAML

File: readme.md (hunk 3 above)
Comment (ebirving) [inline readme.md:159]: :+1:

File: readme.md (hunk 1 above)
Comment (jsm13) [inline readme.md:111]: Students can checkout the lab-starter branch for starter code. The working app.rb is in there with the necessary Gemfile, db/ files and models (everything they would have done if they followed along in class).

**Comment by RobertAKARobin** [comment]
This is awesome. I got nothin' that Jesse didn't already say. Major props.

**Pull request merged**